WalletWasabi/Wallets/CpfpInfoProvider.cs

```diff
@@ -0,0 +1,218 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Nito.AsyncEx;
+using WalletWasabi.Blockchain.Transactions;
+using WalletWasabi.Extensions;
+using WalletWasabi.Logging;
+using WalletWasabi.Models;
+using WalletWasabi.Tor.Http;
+using WalletWasabi.Tor.Http.Extensions;
+using WalletWasabi.WebClients.Wasabi;
+
+namespace WalletWasabi.Wallets;
+
+public class CpfpInfoProvider : BackgroundService
+{
+	private const int MaximumDelayInSeconds = 120;
+	private const int MaximumScheduledRequestsInParallel = 15;
+	private static readonly TimeSpan MinimumBetweenUpdateRequests = TimeSpan.FromSeconds(MaximumDelayInSeconds);
+
+	public CpfpInfoProvider(WasabiHttpClientFactory httpClientFactory, Network network)
+	{
+		if (network == Network.Main)
+		{
+			_httpClient = httpClientFactory.NewHttpClient(() => new Uri("https://mempool.space/api/"), Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+		}
+		else if(network == Network.TestNet)
+		{
+			_httpClient = httpClientFactory.NewHttpClient(() => new Uri("https://mempool.space/testnet/api/"), Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+		}
+		else
+		{
+			throw new InvalidOperationException("CpfpInfoProvider is only operational on Main or TestNet");
+		}
+	}
+
+
+	private readonly IHttpClient _httpClient;
+
+
+	private readonly Channel<SmartTransaction> _transactionsChannel = Channel.CreateUnbounded<SmartTransaction>();
+	private readonly Dictionary<uint256, CachedCpfpInfo> _cpfpInfoCache = new();
+	private AsyncLock AsyncLock { get; } = new();
+
+	private DateTimeOffset _lastUpdateCacheLoop = DateTimeOffset.MinValue;
+
+	public event EventHandler<EventArgs>? RequestedCpfpInfoArrived;
+
+
+	protected override async Task ExecuteAsync(CancellationToken cancel)
+	{
+		List<uint256> scheduledRequests = [];
+		List<Task> tasks = [];
+		while (!cancel.IsCancellationRequested)
+		{
+			var txToFetch = await _transactionsChannel.Reader.ReadAsync(cancel).ConfigureAwait(false);
+			var txid = txToFetch.GetHash();
+
+			using (await AsyncLock.LockAsync(cancel).ConfigureAwait(false))
+			{
+				if (scheduledRequests.Contains(txid) ||
+				    !ShouldRequest(txToFetch) ||
+				    (_cpfpInfoCache.TryGetValue(txid, out var cachedCpfpInfo) &&
+				     cachedCpfpInfo.TimeLastUpdate > DateTimeOffset.UtcNow - MinimumBetweenUpdateRequests))
+				{
+					// We already scheduled fetch or fetching doesn't make sense anymore or last result is recent enough.
+					continue;
+				}
+			}
+
+			scheduledRequests.Add(txid);
+			tasks.Add(Task.Run(() => ScheduleTask(txToFetch), cancel));
+
+			if (tasks.Count <= MaximumScheduledRequestsInParallel)
```

Comment:
This will always remove one task, and only when a new request has to be made. It is wrong because completed tasks should be removed as fast as they finish.

Reply:
I did not 100% fix this but it's better